staging/src/k8s.io/code-generator/cmd/api-linter/linters/api-linter.go

```diff
@@ -0,0 +1,292 @@
+func DefaultNameSystem() string {
+	return "raw"
+}
+
+// Packages returns a API linter that validates API conventions and generates a report file.
+func Packages(context *generator.Context, arguments *args.GeneratorArgs) generator.Packages {
+	return generator.Packages{
+		&generator.DefaultPackage{
+			PackagePath: "k8s.io/kubernetes/staging/src/k8s.io/code-generator/cmd/api-linter/report",
+			GeneratorFunc: func(c *generator.Context) (generators []generator.Generator) {
+				filename := ""
+				if customArgs, ok := arguments.CustomArgs.(*CustomArgs); ok {
+					filename = customArgs.WhitelistFilename
+				}
+				apiValidators := []APIValidator{
+					goJSONNameMatchAPIConvention{},
+				}
+				return []generator.Generator{newAPILinter(arguments.OutputFileBaseName, filename, apiValidators)}
+			},
+			FilterFunc: func(c *generator.Context, t *types.Type) bool {
+				// There is a conflict between this codegen and codecgen, we should avoid types generated for codecgen
+				if strings.HasPrefix(t.Name.Name, "codecSelfer") {
+					return false
+				}
+				pkg := context.Universe.Package(t.Name.Package)
+				if hasAPITagValue(pkg.Comments, tagValueTrue) {
+					return !hasAPITagValue(t.CommentLines, tagValueFalse)
+				}
+				return hasAPITagValue(t.CommentLines, tagValueTrue)
+			},
+		},
+	}
+}
+
+func (l *apiLinter) Filter(c *generator.Context, t *types.Type) bool {
+	// There is a conflict between this codegen and codecgen, we should avoid types generated for codecgen
+	return !strings.HasPrefix(t.Name.Name, "codecSelfer")
+}
+
+func (l *apiLinter) Imports(c *generator.Context) []string {
+	return []string{}
+}
+
+// apiLinter validates API spec against API conventions.
+type apiLinter struct {
+	generator.DefaultGen
+	validators        []APIValidator
+	violations        []apiViolation
+	whitelist         map[apiViolation]bool
+	whitelistFilename string
+}
+
+// apiViolation uniquely identifies one API violation. All three fields (linterName, typeName and violationID) should not contain ','
+// to work properly with whitelist parsing (whitelist entries in the format of: linterName,typeName,violationID).
+type apiViolation struct {
+	// Validator's name from APIValidator.Name()
+	linterName string
+	// Type being validated
+	typeName string
+	// Returned from APIValidator.Validate(t *types.Type). It uniquely identifies one API violation within the validator namespace.
+	violationID string
+}
+
+func newAPILinter(sanitizedName, filename string, apiValidators []APIValidator) generator.Generator {
+	return &apiLinter{
+		DefaultGen: generator.DefaultGen{
+			OptionalName: sanitizedName,
+		},
+		validators:        apiValidators,
+		violations:        []apiViolation{},
+		whitelist:         map[apiViolation]bool{},
+		whitelistFilename: filename,
+	}
+}
+
+func (l *apiLinter) Init(c *generator.Context, w io.Writer) error {
+	if l.whitelistFilename != "" {
+		glog.V(1).Infof("Reading whitelist file: %s", l.whitelistFilename)
```

Comment:
The `generator.Context` is a global context for individual generators to consume. The whitelist / violations file is specific to the openAPIGen / apiLinter generator. IIUC the [generator construction](https://github.com/kubernetes/kube-openapi/blob/bf40560368791a7dddfeea9b3cfcf89b34139f44/pkg/generators/openapi.go#L129) happens after context is constructed. I can move the reading code to where [generator is constructed](https://github.com/kubernetes/kube-openapi/blob/bf40560368791a7dddfeea9b3cfcf89b34139f44/pkg/generators/openapi.go#L164) instead of when generator executes the `Init` method